src/Compiler/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -1,57 +1,123 @@
+    // each time. Instead, we'll hold a ReadOnlyMemory<char> for the string.
+    private readonly LinkedList<ReadOnlyMemory<char>[]> _pages;
+    private int _pageOffset;
+    private char? _lastChar;
+
+    private ReadOnlyMemory<char> _newLine;
+    private int _indentSize;
+    private ReadOnlyMemory<char> _indentString;
 
     private int _absoluteIndex;
     private int _currentLineIndex;
     private int _currentLineCharacterIndex;
 
-    public CodeWriter() : this(Environment.NewLine, RazorCodeGenerationOptions.CreateDefault())
+    public CodeWriter()
+        : this(Environment.NewLine, RazorCodeGenerationOptions.CreateDefault())
     {
     }
 
     public CodeWriter(string newLine, RazorCodeGenerationOptions options)
     {
-        NewLine = newLine;
+        SetNewLine(newLine);
         IndentWithTabs = options.IndentWithTabs;
         TabSize = options.IndentSize;
-        _builder = new StringBuilder();
+
+        _indentSize = 0;
+        _indentString = ReadOnlyMemory<char>.Empty;
+
+        _pages = new();
     }
 
-    public int CurrentIndent { get; set; }
+    private void AddTextChunk(ReadOnlyMemory<char> value)
+    {
+        if (value.Length == 0)
+        {
+            return;
+        }
 
-    public int Length => _builder.Length;
+        // If we're at the start of a page, we need to add the page first.
+        var lastPage = _pageOffset == 0
+            ? _pages.AddLast(new ReadOnlyMemory<char>[PageSize]).Value
+            : _pages.Last.Value;
 
-    public string NewLine
+        // Add our chunk of text (the ReadOnlyMemory<char>) and increment the offset.
+        lastPage[_pageOffset] = value;
+        _pageOffset++;
+
+        // We've reached the end of a page, so we reset the offset to 0.
+        // This will cause a new page to be added next time.
+        if (_pageOffset == PageSize)
+        {
+            _pageOffset = 0;
+        }
+
+        // Remember the last character of the text chunk we just added.
+        _lastChar = value.Span[^1];
+    }
+
+    public int CurrentIndent
     {
-        get => _newLine;
+        get => _indentSize;
         set
         {
-            if (value == null)
+            if (value < 0)
             {
-                throw new ArgumentNullException(nameof(value));
+                throw new ArgumentOutOfRangeException(nameof(value));
             }
 
-            if (value != "\r\n" && value != "\n")
+            if (_indentSize != value)
             {
-                throw new ArgumentException(Resources.FormatCodeWriter_InvalidNewLine(value), nameof(value));
+                _indentSize = value;
+                _indentString = ComputeIndent(value, IndentWithTabs, TabSize);
             }
+        }
+    }
+
+    // Because of the how _absoluteIndex is computed, it is effectively the length
+    // of what has been written.
+    public int Length => _absoluteIndex;
+
+    public string NewLine
+    {
+        get => _newLine.ToString();
```

Comment:
Does `ReadOnlyMemory<char>.ToString()` return the string or is it the default `object.ToString` implementation? Also why not keep `_newLine` to be a `string` since it's never initialized to anything else?

Reply:
You're right though, I'll just switch this over to a string.